File: apps/server/src/main/resources/scripts/javascript/regex-runner.js
const pattern = process.env.PATTERN;
const testString = process.env.TEST_STRING;

const MAX_MATCHES = 1000;
const MAX_MATCH_LEN = 10000;

const start = Date.now();

const response = {
  engine: "javascript",
  pattern,
  success: false,
  matches: [],
  matchCount: 0,
  matchRanges: [],
  groups: [],
  namedGroups: {},
  warnings: [],
  isFullMatch: false,
  error: null,
  executionTimeMs: null,
};

if (!pattern || testString === undefined) {
  response.error =
    "Variáveis de ambiente 'PATTERN' e 'TEST_STRING' são obrigatórias";
  console.error(JSON.stringify(response));
  process.exit(1);
}

try {
  const regex = new RegExp(pattern, "g");

  let count = 0;
  const iterator = testString.matchAll(regex);

  for (const match of iterator) {
    if (count >= MAX_MATCHES) {
      response.warnings.push(`maxMatches truncated to ${MAX_MATCHES}`);
      break;
    }

    let fullMatch = match[0] || "";

    if (fullMatch.length > MAX_MATCH_LEN) {
      fullMatch = fullMatch.substring(0, MAX_MATCH_LEN);
      response.warnings.push("match truncated to MAX_MATCH_LEN");
    }

    response.matches.push(fullMatch);
    response.matchRanges.push({
      start: match.index,
      end: match.index + fullMatch.length,
    });

    const groupList = [];
    for (let i = 0; i < match.length; i++) {
      groupList.push(match[i]);
    }
    response.groups.push(groupList);

    if (match.groups) {
      for (const [key, value] of Object.entries(match.groups)) {
        if (!response.namedGroups[key]) {
          response.namedGroups[key] = [];
        }
        response.namedGroups[key].push(value);
      }
    }

    count++;
  }

  response.matchCount = response.matches.length;

  const fullRegex = new RegExp(`^${pattern}$`);
  response.isFullMatch = fullRegex.test(testString);

  response.success = true;
} catch (e) {
  response.success = false;
  response.error = e.message;
}

response.executionTimeMs = Date.now() - start;

if (response.success) {
  console.log(JSON.stringify(response));
} else {
  console.error(JSON.stringify(response));
  process.exit(1);
}

File: apps/server/src/main/resources/scripts/python/regex_runner.py
#!/usr/bin/env python3
import os, json, re, time, sys

MAX_MATCHES = 1000
MAX_MATCH_LEN = 10000

pattern = os.environ.get('PATTERN')
test_string = os.environ.get('TEST_STRING')

out = {
    "engine": "python",
    "pattern": pattern,
    "success": False,
    "matches": [],
    "matchCount": 0,
    "matchRanges": [],
    "groups": [],
    "namedGroups": {},
    "warnings": [],
    "error": None,
    "executionTimeMs": None
}

start = time.time()
if not pattern or test_string is None:
    out['error'] = "Missing PATTERN or TEST_STRING"
    print(json.dumps(out))
    sys.exit(1)

try:
    compiled = re.compile(pattern)
    count = 0
    for m in compiled.finditer(test_string):
        if count >= MAX_MATCHES:
            out['warnings'].append(f"maxMatches truncated to {MAX_MATCHES}")
            break
        full = m.group(0) or ""
        original_len = len(full)
        if len(full) > MAX_MATCH_LEN:
            full = full[:MAX_MATCH_LEN]
            out['warnings'].append("match truncated to MAX_MATCH_LEN")
        out['matches'].append(full)
        out['matchRanges'].append({"start": m.start(), "end": m.start() + len(full)})
        # groups
        gs = []
        for i in range(0, len(m.groups())+1):
            try:
                g = m.group(i)
            except IndexError:
                g = None
            if g is not None and len(g) > MAX_MATCH_LEN:
                g = g[:MAX_MATCH_LEN]
            gs.append(g)
        out['groups'].append(gs)
        for name, val in m.groupdict().items():
            if name not in out['namedGroups']:
                out['namedGroups'][name] = []
            out['namedGroups'][name].append(val)
        count += 1

    out['matchCount'] = len(out['matches'])
    out['success'] = True
    out['isFullMatch'] = compiled.fullmatch(test_string) is not None

except re.error as e:
    out['error'] = "PatternError: " + str(e)
    out['success'] = False
except Exception as e:
    out['error'] = "RuntimeError: " + str(e)
    out['success'] = False
finally:
    out['executionTimeMs'] = int((time.time() - start) * 1000)
    print(json.dumps(out))